### ISSUE TITLE
Migrate AWI-CIROH to a new GCP project with billing managed by them

### ISSUE THREAD

**Issue by yuvipanda** [2i2c-org/infrastructure#4238]
Note: They did give us more than 6 weeks notice, in a meeting attended by me and @jmunro. Unfortunately while I got access, I didn't actually create an appropriate issue at that point.

AWI-CIROH is moving from a billing account + project managed by us to one managed by them. This is great news, because we no longer have to pass through cloud costs here. 

However, we will have to complete this migration before June 30th.

I've received access to this google cloud project, and invited everyone else to it as well with appropriate permissions.

https://2i2c.freshdesk.com/a/tickets/1687 has more information.

```[tasklist]
- [x] Get all engineers to accept permissions on this GCP project
- [x] Create new infrastructure on this new GCP project: https://github.com/2i2c-org/infrastructure/pull/4262
- [x] Transfer *initial* filestore contents
- [x] Do final migration (involving some downtime)
- [ ] Have the community validate this
```

**Comment by sgibson91** [comment]
Invite accepted!

**Comment by GeorgianaElena** [comment]
I don't see any invite in my email inbox :/ Wondering if it can be resent?

**Comment by sgibson91** [comment]
@GeorgianaElena I have just added you again

**Comment by GeorgianaElena** [comment]
Hmm, still can't see it :( . Was it sent to georgianaelena@2i2c.org?

**Comment by sgibson91** [comment]
@GeorgianaElena Yes, have you checked spam?

**Comment by sgibson91** [comment]
<img width="806" alt="Screenshot 2024-06-18 at 11 01 57" src="https://github.com/2i2c-org/infrastructure/assets/44771837/be0016ec-73a3-458f-b408-668c56077cf9">

I can see you in the IAM panel

Try visiting here? https://console.cloud.google.com/iam-admin/iam?authuser=1&project=awi-ciroh

**Comment by sgibson91** [comment]
Hang on, I might be in the wrong project

**Comment by sgibson91** [comment]
I have Slacked you an invite link directly based on what I received in my email

**Comment by GeorgianaElena** [comment]
Invitation accepted. Thanks @sgibson91

**Comment by sgibson91** [comment]
I'm getting a whole bunch of errors trying to deploy the terraform config to the new AWI-CIROH project.

I think these ones specifically are related to organisational setting:

```
 Error: googleapi: Error 412: Request violates constraint 'constraints/storage.uniformBucketLevelAccess', conditionNotMet
│ 
│   with google_storage_bucket.usage_logs_bucket,
│   on buckets.tf line 47, in resource "google_storage_bucket" "usage_logs_bucket":
│   47: resource "google_storage_bucket" "usage_logs_bucket" {
│ 
╵
╷
│ Error: Error creating service account key: googleapi: Error 400: Key creation is not allowed on this service account.
│ Details:
│ [
│   {
│     "@type": "type.googleapis.com/google.rpc.PreconditionFailure",
│     "violations": [
│       {
│         "description": "Key creation is not allowed on this service account.",
│         "subject": "projects/ciroh-jupyterhub-423218/serviceAccounts/awi-ciroh-cd-sa@ciroh-jupyterhub-423218.iam.gserviceaccount.com?configvalue=awi-ciroh-cd-sa%40ciroh-jupyterhub-423218.iam.gserviceaccount.com",
│         "type": "constraints/iam.disableServiceAccountKeyCreation"
│       }
│     ]
│   }
│ ]
│ , failedPrecondition
│ 
│   with google_service_account_key.cd_sa,
│   on cd.tf line 23, in resource "google_service_account_key" "cd_sa":
│   23: resource "google_service_account_key" "cd_sa" {
```

**Comment by sgibson91** [comment]
The rest of the errors are as follows:

```
Error: googleapi: Error 400: Project "ciroh-jupyterhub-423218" has no network named "default".
│ Details:
│ [
│   {
│     "@type": "type.googleapis.com/google.rpc.RequestInfo",
│     "requestId": "0x4de250f6194a4030"
│   }
│ ]
│ , badRequest
│ 
│   with google_container_cluster.cluster,
│   on cluster.tf line 48, in resource "google_container_cluster" "cluster":
│   48: resource "google_container_cluster" "cluster" {
│ 
╵
╷
│ Error: Error creating Instance: googleapi: Error 400: network 'default' does not exist.
│ 
│   with google_filestore_instance.homedirs[0],
│   on storage.tf line 1, in resource "google_filestore_instance" "homedirs":
│    1: resource "google_filestore_instance" "homedirs" {
│ 
╵
╷
│ Error: Error applying IAM policy for service account 'projects/ciroh-jupyterhub-423218/serviceAccounts/awi-ciroh-staging@ciroh-jupyterhub-423218.iam.gserviceaccount.com': Error setting IAM policy for service account 'projects/ciroh-jupyterhub-423218/serviceAccounts/awi-ciroh-staging@ciroh-jupyterhub-423218.iam.gserviceaccount.com': googleapi: Error 400: Identity Pool does not exist (ciroh-jupyterhub-423218.svc.id.goog). Please check that you specified a valid resource name as returned in the `name` attribute in the configuration API., badRequest
│ 
│   with google_service_account_iam_binding.workload_identity_binding["staging"],
│   on workload-identity.tf line 26, in resource "google_service_account_iam_binding" "workload_identity_binding":
│   26: resource "google_service_account_iam_binding" "workload_identity_binding" {
│ 
╵
╷
│ Error: Error applying IAM policy for service account 'projects/ciroh-jupyterhub-423218/serviceAccounts/awi-ciroh-prod@ciroh-jupyterhub-423218.iam.gserviceaccount.com': Error setting IAM policy for service account 'projects/ciroh-jupyterhub-423218/serviceAccounts/awi-ciroh-prod@ciroh-jupyterhub-423218.iam.gserviceaccount.com': googleapi: Error 400: Identity Pool does not exist (ciroh-jupyterhub-423218.svc.id.goog). Please check that you specified a valid resource name as returned in the `name` attribute in the configuration API., badRequest
│ 
│   with google_service_account_iam_binding.workload_identity_binding["prod"],
│   on workload-identity.tf line 26, in resource "google_service_account_iam_binding" "workload_identity_binding":
│   26: resource "google_service_account_iam_binding" "workload_identity_binding" {
```

Possibly due to certain resources being created yet?

**Comment by sgibson91** [comment]
I cannot override the Bucket Level Access, so we will have to debug that

<img width="674" alt="Screenshot 2024-06-20 at 11 09 42" src="https://github.com/2i2c-org/infrastructure/assets/44771837/15439200-fdfa-4eca-82b2-b1b0bae21781">

**Comment by sgibson91** [comment]
I think GCP has stopped providing a default network so we shall have to create one

**Comment by sgibson91** [comment]
I had to bump the Basic HDD (Standard) capacity (GB) per region Quota to allow for the size of the existing filestore

**Comment by sgibson91** [comment]
Service account key creation is an organisational policy too

<img width="954" alt="Screenshot 2024-06-20 at 13 59 16" src="https://github.com/2i2c-org/infrastructure/assets/44771837/1d4ab660-24ab-4a9c-a11f-c6c7a10ab37e">

**Comment by sgibson91** [comment]
I have managed to deploy a cluster and filestore, so we should be unblocked to do the file transfer. Solving the service account key and bucket access problems will require either: 1) discussion with AWI-CIROH team to relax restrictions, or 2) further debugging in terraform to work around these restrictions. The service account key creation part is particularly critical as we won't be able to deploy anything using the deployer or in CI/CD until we have a key.

**Comment by sgibson91** [comment]
I have hubs up

- https://staging.ciroh2.awi.2i2c.cloud
- https://ciroh2.awi.2i2c.cloud

**Comment by yuvipanda** [comment]
Following up on our conversation, I've communicated to the community that they need to relax the serviceaccount restriction.

Let's also just turn off all object storage, and I'll communicate that to them as well.

**Comment by sgibson91** [comment]
I'm trying to redeploy the hubs buf NFS mounting is failing for the share creator job and the prometheus shared dirsize and volume metric pods. Connection just keeps timing out, so I'm guessing something isn't quite right in the network.

**Comment by yuvipanda** [comment]
We ran into another issue that prevented us from being able to ssh into anything in this new project. I'm working with AWI-CIROH folks to fix that, and so we can start our copy.

I did experiment with figuring out how we can copy filestores across projects. I chose to use a wireguard VPN, and the following pieces of documentation have helped me accomplish this:

1. https://www.digitalocean.com/community/tutorials/how-to-set-up-wireguard-on-ubuntu-22-04
2. https://docs.pi-hole.net/guides/vpn/wireguard/internal/

I set up a test VM on the shared 2i2c project to validate this, and i can access both filestores via this mechanism!

Once the ssh issues are sorted out, I'll start the copy.

Note that this *will* be charged by google at 0.01$ / GB of network transfer. However, at 2TB only, that's ~20$ so we're ok

**Comment by sgibson91** [comment]
I am not sure what this means for our GPUs on this project 🤔 

> Your quota request for ciroh-jupyterhub-423218 has been partially approved and your project quota has been adjusted according to the following requested limits:
> 
> +------------------+------------+--------+-----------------+----------------+
> | NAME             | DIMENSIONS | REGION | REQUESTED LIMIT | APPROVED LIMIT |
> +------------------+------------+--------+-----------------+----------------+
> | GPUS_ALL_REGIONS |            | GLOBAL |               4 |              4 |
> +------------------+------------+--------+-----------------+----------------+
> 
> Unfortunately, we were unable to grant your below quota request(s):
> 
> +----------------+--------------------+-------------+
> | NAME           | DIMENSIONS         | REGION      |
> +----------------+--------------------+-------------+
> | NVIDIA_T4_GPUS | region=us-central1 | us-central1 |
> +----------------+--------------------+-------------+

From the dashboard it looks like we are still stuck at 1 in the `us-central1` region. Maybe it's one of those things that after the community starts using it, they'll increase the quota?

**Comment by sgibson91** [comment]
Hmmm, there may be something up with the taints as I can't spin up a server on the new staging hub

> 2024-06-24T13:38:56Z [Normal] pod didn't trigger scale-up: 3 node(s) didn't match Pod's node affinity/selector, 1 node(s) had untolerated taint {k8s.dask.org_dedicated: worker}, 1 node(s) had untolerated taint {nvidia.com/gpu: present}, 1 in backoff after failed scale-up

**Comment by yuvipanda** [comment]
I've started a copy process, and me and @sgibson91 have scheduled time tomorrow to make progress on the taints situation

**Comment by yuvipanda** [comment]
Had a quick session with @sgibson91 and she is unblocked now.

**Comment by sgibson91** [comment]
We established the failure of the server to start was due to exceeded quota, but also the small server was being scheduled to a n2-highmem-16 node instead of a smaller -4 node. I'm going to see if this issue is also present on the old hub (since we copied config, the likelihood is high) so I can debug that there without the quota issues. I will also incrementally bump the CPU quota in the new project until it matches the old.

**Comment by sgibson91** [comment]
This is the only present taint on the listed node

```
Taints:             hub.jupyter.org_dedicated=user:NoSchedule
```

**Comment by sgibson91** [comment]
Pods are being scheduled on n2-highmem-16 node types because that is what the node selector for _every_ profile in the list (in both staging and prod hubs) is set to

- staging: https://github.com/2i2c-org/infrastructure/blob/096c218e9748c3516d378327bc850deb96faf300/config/clusters/awi-ciroh/staging.values.yaml#L65
- prod: https://github.com/2i2c-org/infrastructure/blob/096c218e9748c3516d378327bc850deb96faf300/config/clusters/awi-ciroh/prod.values.yaml#L88

Was this intentional or is it a typo?

**Comment by sgibson91** [comment]
I think setting all the node selectors to the -16 nodes was set in [this PR](https://github.com/2i2c-org/infrastructure/pull/2539) and never properly reverted. There's a FIXME note that links to this PR, but I don't really understand it or what the action is.

**Comment by sgibson91** [comment]
The CPU quota is now at 48, which allows servers to start. I think we just need to make a decision on what the node selector values should be for the profiles and then I'm done.

**Comment by yuvipanda** [comment]
@sgibson91 ah, I see - if it was in that setup earlier, let's just let that be. We can change that later on. And yes, please do continue asking for increases.

**Comment by yuvipanda** [comment]
Unfortunately I had to abort because we don't have enough quota, so this won't be done today. I've communicated this to the community.

**Comment by sgibson91** [comment]
I bumped the CPU quota again just now and got another weird "partially approved" response where the quota for all regions has been approved, but not for the us-central1 region.

```
Your quota request for ciroh-jupyterhub-423218 has been partially approved and your project quota has been adjusted according to the following requested limits:

+------------------+------------+--------+-----------------+----------------+
| NAME             | DIMENSIONS | REGION | REQUESTED LIMIT | APPROVED LIMIT |
+------------------+------------+--------+-----------------+----------------+
| CPUS_ALL_REGIONS |            | GLOBAL |             128 |            128 |
+------------------+------------+--------+-----------------+----------------+

Unfortunately, we were unable to grant your below quota request(s):

+---------+--------------------+-------------+
| NAME    | DIMENSIONS         | REGION      |
+---------+--------------------+-------------+
| N2_CPUS | region=us-central1 | us-central1 |
+---------+--------------------+-------------+
```

This may be a case where we let the community know and they can talk directly with their Google rep to get the appropriate quotas put in place.

**Comment by yuvipanda** [comment]
@sgibson91 yeah i've asked them to escalate via their service rep.

**Comment by haroldcampbell** [comment]
@yuvipanda are we waiting on the community rep?

**Comment by yuvipanda** [comment]
@haroldcampbell yes.

**Comment by yuvipanda** [comment]
This has been unblocked, and wednesday evening pacific has been set as migration time.

**Comment by yuvipanda** [comment]
I migrated the following bits of data:

- [x] Home directories
- [x] Prometheus data
- [x] Hub DB dirs

**Comment by yuvipanda** [comment]
This is finally complete, oof. Let's do a postmortem soon.
